L1Trigger/L1CaloTrigger/interface/Phase2L1CaloEGammaEmulator.h

```diff
@@ -0,0 +1,1738 @@
+
+  /* 
+   * card class: represents one RCT card. Each card has five 3x4 regions and one 2x4 region,
+   *             which is represented by a 3x4 region with its third row zero'd out.
+   *             idx 0-35: odd values of cardIdx span eta = 0 to eta = 1.41 
+   *                       even values of cardIdx span eta = -1.41 to eta = 0
+   *             The realEta and realPhi arrays store the (eta, phi) of the center of the towers.
+   */
+
+  class card {
+  private:
+    int idx_ = -1;
+    region3x4 card3x4Regions[N_REGIONS_PER_CARD];
+    towers3x4 card3x4Towers[N_REGIONS_PER_CARD];
+
+  public:
+    // constructor
+    card() {
+      idx_ = -1;
+      for (int i = 0; i < N_REGIONS_PER_CARD; i++) {
+        card3x4Regions[i].setIdx(i);
+        card3x4Regions[i].zeroOut();
+        card3x4Towers[i].setIdx(i);
+        card3x4Towers[i].zeroOut();
+      }
+    };
+
+    // copy constructor
+    card(const card& other) {
+      idx_ = other.idx_;
+      for (int i = 0; i < N_REGIONS_PER_CARD; i++) {
+        card3x4Regions[i] = other.card3x4Regions[i];
+        card3x4Towers[i] = other.card3x4Towers[i];
+      }
+    };
+
+    // overload operator= to use copy constructor
+    card operator=(const card& other) {
+      const card& newCard(other);
+      return newCard;
+    };
+
+    // set members
+    inline void setIdx(int idx) { idx_ = idx; };
+    inline void zeroOut() {
+      for (int i = 0; i < N_REGIONS_PER_CARD; i++) {
+        card3x4Regions[i].zeroOut();
+        card3x4Towers[i].zeroOut();
+      };
+    };
+
+    // get members
+    inline float getIdx() const { return idx_; };
+    inline region3x4& getRegion3x4(int idx) {
+      assert(idx < N_REGIONS_PER_CARD);
+      return card3x4Regions[idx];
+    }
+    inline towers3x4& getTowers3x4(int idx) {
+      assert(idx < N_REGIONS_PER_CARD);
+      return card3x4Towers[idx];
+    }
+  };
+
+  /*******************************************************************/
+
+  /*
+   *  Crystal class for RCT
+   */
+
+  class crystal {
+  public:
+    ap_uint<10> energy;
+
+    crystal() {
+      energy = 0;
+      //    timing = 0;
+    }
+
+    crystal(ap_uint<10> energy) {  // To-do: add timing information
+      this->energy = energy;
+      //    this->timing = 0;
+    }
+
+    crystal& operator=(const crystal& rhs) {
+      energy = rhs.energy;
+      //    timing = rhs.timing;
+      return *this;
+    }
+  };
+
+  /*
+  * crystalMax class for RCT
+  */
+  class crystalMax {
+  public:
+    ap_uint<10> energy;
+    uint8_t phiMax;
+    uint8_t etaMax;
+
+    crystalMax() {
+      energy = 0;
+      phiMax = 0;
+      etaMax = 0;
+    }
+
+    crystalMax& operator=(const crystalMax& rhs) {
+      energy = rhs.energy;
+      phiMax = rhs.phiMax;
+      etaMax = rhs.etaMax;
+      return *this;
+    }
+  };
+
+  class ecaltp_t {
+  public:
+    ap_uint<10> energy;
+    ap_uint<5> phi;
+    ap_uint<5> eta;
+  };
+
+  class etaStrip_t {
+  public:
+    ecaltp_t cr0;
```

Comment:
These `crX` here, the `etaStripX` (in `ecalRegion_t`) and the `pkX` may be transferred to an array-like structure? I'm asking because in the `p2eg::initStructure`, `p2eg::getPeakBin20N`, `p2eg::getPeakBin15N`, `p2eg::getClusterPosition` and `p2eg::getECALTowersEt` are quite cumbersome because of this choice (seems to me they could be simplified with a loop).

Reply:
Hi @AdrianoDee, these parts of the code came from the firmware; our plan is to make minimal changes to the firmware blocks to preserve the ease of translating firmware back and forth.

Reply:
Fine by me, thanks for the clarification.